src/env.rs

```diff
@@ -0,0 +1,227 @@
+        let ExportItem::Function(func) = env_ptr_export.item else {
+            panic!()
+        };
+        let FunctionKind::Local(local_func) = &module.funcs.get(func).kind else {
+            panic!()
+        };
+        let func_body = local_func.block(local_func.entry_block());
+        if func_body.instrs.len() != 1 {
+            return Err(anyhow!(
+                "Unexpected get_env_ptr implementation. Should be a constant address return."
+            ));
+        }
+        let Instr::Const(Const {
+            value: Value::I32(env_ptr_addr),
+        }) = &func_body.instrs[0].0
+        else {
+            return Err(anyhow!(
+                "Unexpected get_env_ptr implementation. Should be a constant address return."
+            ));
+        };
+        *env_ptr_addr as u32
+    };
+
+    // If host env is disabled, remove its import entirely
+    // replacing it with a stub panic
+    if matches!(env.host, HostEnv::None) {
+        stub_imported_func(module, "wasi:cli-base/environment", "get-environment");
+        // we do arguments as well because virt assumes reactors for now...
+        stub_imported_func(module, "wasi:cli-base/environment", "get-arguments");
+    }
+
+    let memory = module.memories.iter().nth(0).unwrap().id();
+
+    // prepare the field data list vector for writing
+    // strings must be sorted as binary searches are used against this data
+    let mut field_data_vec: Vec<&str> = Vec::new();
+    for (key, value) in &env.overrides {
+        field_data_vec.push(key.as_ref());
+        field_data_vec.push(value.as_ref());
+    }
+    field_data_vec.sort();
+    match &env.host {
+        HostEnv::Allow(allow_list) => {
+            let mut allow_list: Vec<&str> = allow_list.iter().map(|item| item.as_ref()).collect();
+            allow_list.sort();
+            for key in allow_list {
+                field_data_vec.push(key);
+            }
+        }
+        HostEnv::Deny(deny_list) => {
+            let mut deny_list: Vec<&str> = deny_list.iter().map(|item| item.as_ref()).collect();
+            deny_list.sort();
+            for key in deny_list {
+                field_data_vec.push(key);
+            }
+        }
+        _ => {}
+    }
+
+    let mut field_data_bytes = Vec::new();
+    for str in field_data_vec {
+        assert!(field_data_bytes.len() % 4 == 0);
+        // write the length at the aligned offset
+        field_data_bytes.extend_from_slice(&(str.len() as u32).to_le_bytes());
+        let str_bytes = str.as_bytes();
+        field_data_bytes.extend_from_slice(str_bytes);
+        let rem = str_bytes.len() % 4;
+        // add padding for alignment if necessary
+        if rem > 0 {
+            field_data_bytes.extend((0..4 - rem).map(|_| 0));
+        }
+    }
+
+    let field_data_addr = if field_data_bytes.len() > 0 {
+        // Offset the stack global by the static field data length
+        let field_data_addr = bump_stack_global(module, field_data_bytes.len() as i32)?;
```

Comment:
In the future, as we need to use passive data segments for virtualizing file reads anyway, it might make sense to change this over to creating a new passive data segment that stores the environment data that the adapter can (somehow) read from on-demand rather than "stealing" space from the shadow stack.